File: clojure/src/clj/kmeans/algo.clj
(ns kmeans.algo)

(defn sq ^double [^double x] (* x x))

(defn v- [[^double a ^double b] [^double c ^double d]]
  [(- a c) (- b d)])

(defn v+ [[^double a ^double b] [^double c ^double d]]
  [(+ a c) (+ b d)])

(defn vdiv [[^double a ^double b] ^double k]
  [(/ a k) (/ b k)])

(defn norm [[^double a ^double b]]
  (Math/sqrt (+ (sq a) (sq b))))

(defn dist [v w] (norm (v- v w)))

(defn closest [x choices]
  (apply min-key #(dist x %) choices))

(defn avg [xs]
  (vdiv
    (reduce v+ xs)
    (count xs)))

(defn clusters [xs centroids]
  (vals (group-by #(closest % centroids) xs)))

(defn run [points n iters]
  (let
    [centroids-seq (iterate
      (fn [centroids]
        (map avg (clusters points centroids)))
      (take n points))
    final-centroids (nth centroids-seq iters)]
    (clusters points final-centroids)))

File: scala-js/src/main/scala/kmeans/Algo.scala
package kmeans

import math.sqrt


object Algo {
  val n = 10
  val iters = 15

  class Point(val x: Double, val y: Double) {
    def /(k: Double): Point = new Point(x / k, y / k)

    def +(p: Point) = new Point(x + p.x, y + p.y)
    def -(p: Point) = new Point(x - p.x, y - p.y)

    def modulus = sqrt(sq(x) + sq(y))
  }

  def run(xs: List[Point]) = {
    var centroids = xs take n

    for (i <- 1 to iters) {
      centroids = clusters(xs, centroids) map average
    }
    clusters(xs, centroids)
  }

  def clusters(xs: List[Point], centroids: List[Point]) =
    (xs groupBy { x => closest(x, centroids) }).values.toList

  def closest(x: Point, choices: List[Point]) =
    choices minBy { y => dist(x, y) }

  def sq(x: Double) = x * x

  def dist(x: Point, y: Point) = (x - y).modulus

  def average(xs: List[Point]) = xs.reduce(_ + _) / xs.size
}